src/main/java/sg/ncl/MainController.java

```diff
@@ -2847,14 +2856,50 @@ public String internetRequest(@PathVariable String teamName,
     }
 
     @RequestMapping("/web_ssh/access_node/{qualifiedName:.+}")
-    public String webAccessNode(Model model, HttpSession session, @PathVariable String qualifiedName) throws WebServiceRuntimeException {
+    public String sshAccessNode(Model model, HttpSession session, @PathVariable String qualifiedName) throws WebServiceRuntimeException {
         getDeterUid(model, session);
         model.addAttribute("qualified", qualifiedName);
         model.addAttribute("cols", ptyProperties.getCols());
         model.addAttribute("rows", ptyProperties.getRows());
         return "webssh";
     }
 
+    @RequestMapping("/web_vnc/access_node/{qualifiedName:.+}/{portnum}")
+    public String vncAccessNode(@PathVariable String qualifiedName, @PathVariable String portnum) throws NoSuchAlgorithmException {
+        UriComponents uriComponents = UriComponentsBuilder.fromUriString(vncProperties.getHttp())
+                .queryParam("host", vncProperties.getHost())
+                .queryParam("port", vncProperties.getPort())
+                .queryParam("path", qencode(qualifiedName + ":" + portnum))
+                .build();
+        log.info("VNC URI: {}", uriComponents.toString());
+        return "redirect:" + uriComponents.toString();
+    }
+
+    // Reference: http://www.baeldung.com/sha-256-hashing-java
+    private String bytesToHex(byte[] hash) {
+        StringBuffer hexString = new StringBuffer();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Replace the synchronized class "StringBuffer" by an unsynchronized one such as "StringBuilder". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1149)